polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -193,121 +150,18 @@ impl Artifacts {
 	///
 	/// Create the cache directory on-disk if it doesn't exist.
-	pub async fn new_and_prune(cache_path: &Path) -> Self {
-		let mut artifacts = Self { inner: HashMap::new() };
-		let _ = artifacts.insert_and_prune(cache_path).await.map_err(|err| {
-			gum::error!(
-				target: LOG_TARGET,
-				"could not initialize artifacts cache: {err}",
-			)
-		});
-		artifacts
-	}
-
-	async fn insert_and_prune(&mut self, cache_path: &Path) -> Result<(), String> {
-		async fn is_corrupted(path: &Path) -> bool {
-			let checksum = match tokio::fs::read(path).await {
-				Ok(bytes) => blake3::hash(&bytes),
-				Err(err) => {
-					// just remove the file if we cannot read it
-					gum::warn!(
-						target: LOG_TARGET,
-						?err,
-						"unable to read artifact {:?} when checking integrity, removing...",
-						path,
-					);
-					return true
-				},
-			};
-
-			if let Some(file_name) = path.file_name() {
-				if let Some(file_name) = file_name.to_str() {
-					return !file_name.ends_with(checksum.to_hex().as_str())
-				}
-			}
-			true
-		}
-
-		// Insert the entry into the artifacts table if it is valid.
-		// Otherwise, prune it.
-		async fn insert_or_prune(
-			artifacts: &mut Artifacts,
-			entry: &tokio::fs::DirEntry,
-			cache_path: &Path,
-		) -> Result<(), String> {
-			let file_type = entry.file_type().await;
-			let file_name = entry.file_name();
-
-			match file_type {
-				Ok(file_type) =>
-					if !file_type.is_file() {
-						return Ok(())
-					},
-				Err(err) => return Err(format!("unable to get file type for {file_name:?}: {err}")),
-			}
-
-			if let Some(file_name) = file_name.to_str() {
-				let id = ArtifactId::from_file_name(file_name);
-				let path = cache_path.join(file_name);
-
-				if id.is_none() || is_corrupted(&path).await {
-					let _ = tokio::fs::remove_file(&path).await;
-					return Err(format!("invalid artifact {path:?}, file deleted"))
-				}
-
-				let id = id.expect("checked is_none() above; qed");
-				gum::debug!(
-					target: LOG_TARGET,
-					"reusing existing {:?} for node version v{}",
-					&path,
-					NODE_VERSION,
-				);
-				artifacts.insert_prepared(id, path, SystemTime::now(), Default::default());
-
-				Ok(())
-			} else {
-				Err(format!("non-Unicode file name {file_name:?} found in {cache_path:?}"))
-			}
-		}
-
+	pub async fn new(cache_path: &Path) -> Self {
 		// Make sure that the cache path directory and all its parents are created.
-		if let Err(err) = tokio::fs::create_dir_all(cache_path).await {
-			if err.kind() != io::ErrorKind::AlreadyExists {
-				return Err(format!("failed to create dir {cache_path:?}: {err}"))
-			}
-		}
-
-		let mut dir = tokio::fs::read_dir(cache_path)
-			.await
-			.map_err(|err| format!("failed to read dir {cache_path:?}: {err}"))?;
+		// First delete the entire cache. Nodes are long-running so this should populate shortly.
+		let _ = tokio::fs::remove_dir_all(cache_path).await;
```

Comment:
Instead of `remove_dir_all` could we have a preset unique file extension for the cached files (e.g. maybe `.pvfcache`) and then over a directory's contents and only delete files with that given extension? That should only be a couple of extra lines but would be a lot safer if something goes wrong with where the `cache_path` points to.

`remove_dir_all` always makes me nervous because it's essentially an `rm -Rf`, so if for some reason `cache_path` would be something that also contains other stuff (e.g. in the worst case someone would want to have a cache in their `/home/$user` :P) this would also unnecessarily delete all of it.

Reply:
Another very sensible suggestion. :) Should we try to clean up any existing artifacts with the current naming scheme? e.g. delete any files starting with `wasmtime_` which would be compatible with the [current](https://github.com/paritytech/polkadot-sdk/pull/1918/files#diff-8b70b8fcbc7a957f4e0cca743522649ffc8d2ee24ef49531e8318392bdffddbbR74) scheme as well as the [one before artifact persistence](https://github.com/paritytech/polkadot-sdk/pull/1918/files#diff-8b70b8fcbc7a957f4e0cca743522649ffc8d2ee24ef49531e8318392bdffddbbL78). We also keep the self-contained "worker dirs" in there, which should be deleted:

![Screenshot 2024-01-10 at 08 53 59](https://github.com/paritytech/polkadot-sdk/assets/6035856/c6ecb5e6-bcba-4176-aa7a-98785e7d136b)

Maybe not ideal, but indeed seems better than nuking the whole directory.